appbuilder/services/npm-service.ts

```diff
@@ -0,0 +1,233 @@
+		private $hostInfo: IHostInfo,
+		private $httpClient: Server.IHttpClient,
+		private $logger: ILogger,
+		private $projectConstants: Project.IConstants) { }
+
+	@exportedPromise("npmService")
+	public install(projectDir: string, dependencyToInstall?: INpmDependency): IFuture<INpmInstallResult> {
+		return (() => {
+			let npmInstallResult: INpmInstallResult = {};
+
+			if (dependencyToInstall) {
+				npmInstallResult.result = {
+					isInstalled: false,
+					isTypesInstalled: false
+				};;
+
+				try {
+					this.npmInstall(projectDir, dependencyToInstall.name, dependencyToInstall.version, ["--save", "--save-exact"]).wait();
+					npmInstallResult.result.isInstalled = true;
+				} catch (err) {
+					npmInstallResult.error = err;
+				}
+
+				if (dependencyToInstall.installTypes && npmInstallResult.result.isInstalled && this.hasTypesForDependency(dependencyToInstall.name).wait()) {
+					try {
+						this.installTypingsForDependency(projectDir, dependencyToInstall.name).wait();
+						npmInstallResult.result.isTypesInstalled = true;
+					} catch (err) {
+						npmInstallResult.error = err;
+					}
+				}
+			} else {
+				try {
+					this.npmPrune(projectDir).wait();
+					this.npmInstall(projectDir).wait();
+				} catch (err) {
+					npmInstallResult.error = err;
+				}
+			}
+
+			this.generateReferencesFile(projectDir).wait();
+
+			return npmInstallResult;
+		}).future<INpmInstallResult>()();
+	}
+
+	@exportedPromise("npmService")
+	public uninstall(projectDir: string, dependency: string): IFuture<void> {
+		return (() => {
+			let packageJsonContent = this.getPackageJsonContent(projectDir).wait();
+
+			if (packageJsonContent && packageJsonContent.dependencies && packageJsonContent.dependencies[dependency]) {
+				this.npmUninstall(projectDir, dependency, ["--save"]).wait();
+			}
+
+			if (packageJsonContent && packageJsonContent.devDependencies && packageJsonContent.devDependencies[`${NpmService.TYPES_DIRECTORY}${dependency}`]) {
+				this.npmUninstall(projectDir, `${NpmService.TYPES_DIRECTORY}${dependency}`, ["--save-dev"]).wait();
+			}
+		}).future<void>()();
+	}
+
+	public getPackageJsonFromNpmRegistry(packageName: string, version?: string): IFuture<any> {
+		return (() => {
+			let packageJsonContent: any;
+			version = version || "latest";
+			try {
+				let url = this.buildNpmRegistryUrl(packageName, version);
+				// This call will return error with message '{}' in case there's no such package.
+				let result = this.$httpClient.httpRequest(url).wait().body;
+				packageJsonContent = JSON.parse(result);
+			} catch (err) {
+				this.$logger.trace("Error caught while checking the NPM Registry for plugin with id: %s", packageName);
+				this.$logger.trace(err.message);
+			}
+
+			return packageJsonContent;
+		}).future<any>()();
+	}
+
+	private hasTypesForDependency(packageName: string): IFuture<boolean> {
+		return (() => {
+			return !!this.getPackageJsonFromNpmRegistry(`${NpmService.TYPES_DIRECTORY}${packageName}`).wait();
```

Comment:
In the current implementation of `getPackageJsonFromNpmRegistry` is planned to have case in which the method will return `{ }` (it is planned because you expect the request to return {} if the plugin is not found but actually the method will never return `{ }` if you don't address my comment for `this.$httpClient.httpRequest`). If  `getPackageJsonFromNpmRegistry` return `{ }` the result from `hasTypesForDependency` will be true (`!!{ }` is equal to `true`).